src/main/java/apoc/meta/tablesforlabels/PropertyContainerProfile.java

```diff
@@ -0,0 +1,98 @@
+package apoc.meta.tablesforlabels;
+
+import org.neo4j.graphdb.Label;
+import org.neo4j.graphdb.Node;
+import org.neo4j.graphdb.PropertyContainer;
+import org.neo4j.graphdb.schema.ConstraintDefinition;
+
+import java.util.*;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * A profile of a particular ordered label set (or relationship type) has a set of possible properties that can exist, and
+ * stats about those properties.
+ */
+public class PropertyContainerProfile {
+    public long observations;
+    Map<String, PropertyTracker> profile;
+
+    public PropertyContainerProfile() {
+        observations = 0;
+        profile = new HashMap<>(3);
+    }
+
+    public Set<String> propertyNames() { return profile.keySet(); }
+    public PropertyTracker trackerFor(String propName) { return profile.get(propName); }
+
+    public void observe(PropertyContainer n, Iterable<ConstraintDefinition> constraints, boolean isNode, Map<String, Iterable<ConstraintDefinition>> relConstraints, Iterable<Label> relStartNode, Iterable<Label> relEndNode) {
+        observations++;
+
+        for (String propName : n.getPropertyKeys()) {
+            PropertyTracker tracker;
+
+            if (profile.containsKey(propName)) {
+                tracker = profile.get(propName);
+            } else {
+                tracker = new PropertyTracker(propName);
+                profile.put(propName, tracker);
+            }
+
+            tracker.addObservation(n.getProperty(propName));
+
+            if (isNode) {
+
+                // Check for node constraints
+
+                tracker.mandatory = false;
+                for (ConstraintDefinition cd : constraints) {
+                    for (String pk : cd.getPropertyKeys()) {
+                        if (pk == propName) {
+                            tracker.mandatory = true;
+                        }
+                    }
+                }
+            } else {
+
+                // Check for relationship constraints - NOTE: Could probably improve the efficiency here a bit. Too many nested loops.
+
+                List<String> sourceNodeLabels = new ArrayList<String>();
```

Comment:
sourceNodeLabels and targetNodeLabels can be initialized with an argument 4 for initial size.  99.9999% of neo4j nodes don't have more than 4 labels for reasons we can discuss later if you're interested.

Also, you can save a lot of code and two while loops here by initializing your lists from your iterators in an easier way here:  

https://www.baeldung.com/java-convert-iterator-to-list#java8

Reply:
This was entirely re-written in a previous update, so may not be relevant any more..